### ISSUE TITLE
There's no way to reuse a form dialog

### ISSUE THREAD

**Issue by v-kydela** [BotBuilderCommunity/botbuilder-community-dotnet#157]
This issue is very closely related to this one: https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/issues/146

Michael's issue only mentions the private `_formState` field, but there is also the private `_state` field to consider. FormFlow has multiple ways of keeping track of its state and none of them seem to get reset when a form dialog begins or ends. Since the state doesn't get reset automatically and there's no way to do it manually (because it's private), a form dialog cannot be effectively reused. This is especially problematic in Bot Builder v4 because even if you reconstruct the dialog you can't add a dialog to a dialog set if its ID already exists: https://github.com/microsoft/botbuilder-dotnet/issues/2548

There are a few avenues to explore when it comes to fixing this:

- Always automatically reset a form dialog's state when the dialog starts and/or ends
- Automatically reset a form dialog's state conditionally, perhaps depending on a setting that gets passed into the constructor
- Allow manipulation of a form dialog's state from outside the dialog

**Comment by v-kydela** [comment]
Now that I think about it, why doesn't a form dialog persist its state using the `State` property of the dialog instance the same way other dialogs do?

- [Component dialog](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Dialogs/ComponentDialog.cs#L371)
- [Prompt](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/Prompt.cs#L85)

**Comment by garypretty** [comment]
Closing this as it overlaps with #146. @v-kydela per your comment regarding how FormFlow is implemented in terms of state, I think this is likely to do with when it was developed, before GA of v4 if I remember rightly, which means some of the patterns used now were not baked yet. Do you think there is a case to update this now? @EricDahlvang do you have an opinion here?

**Comment by v-kydela** [comment]
@garypretty - I suspect the problems have more to do with trying to follow the patterns of v3 than trying to follow the patterns of v4. Yes, I do still think there is a case to fix this FormFlow package if we want to tell people it's usable. And I did acknowledge the overlap with #146 in my first sentence but I went on to explain why I opened this as its own issue, but it's fine if you want there to be just one issue.
